j5/backends/hardware/sb/arduino.py

```diff
@@ -0,0 +1,281 @@
+        super(SBArduinoHardwareBackend, self).__init__(
+            serial_port=serial_port,
+            serial_class=serial_class,
+            baud=115200,
+        )
+
+        self._digital_pins: Mapping[int, DigitalPinData] = {
+            i: DigitalPinData(mode=GPIOPinMode.DIGITAL_INPUT, state=False)
+            for i in range(2, FIRST_ANALOGUE_PIN)
+        }
+
+        count = 0
+        line = self.read_serial_line(empty=True)
+        while len(line) == 0:
+            line = self.read_serial_line(empty=True)
+            count += 1
+            if count > 25:
+                raise CommunicationError(f"Arduino ({serial_port}) is not responding.")
+
+        if line != "# Booted":
+            raise CommunicationError("Arduino Boot Error.")
+
+        self._version_line = self.read_serial_line()
+
+        if self.firmware_version is not None:
+            version_ids = tuple(map(int, self.firmware_version.split(".")))
+        else:
+            version_ids = (0, 0, 0)
+
+        if version_ids < (2019, 6, 0):
+            raise CommunicationError(
+                f"Unexpected firmware version: {self.firmware_version},",
+                f" expected at least: \"2019.6.0\".",
+            )
+
+        for pin_number in self._digital_pins.keys():
+            self.set_gpio_pin_mode(pin_number, GPIOPinMode.DIGITAL_INPUT)
+
+    @property
+    def firmware_version(self) -> Optional[str]:
+        """The firmware version of the board."""
+        return self._version_line.split("v")[1]
+
+    @handle_serial_error
+    def _command(self, command: str, *params: str) -> List[str]:
+        """Send a command to the board."""
+        message = " ".join([command] + list(params)) + "\n"
+        self._serial.write(message.encode("utf-8"))
+
+        results: List[str] = []
+        while True:
+            line = self.read_serial_line(empty=False)
+            code, param = line.split(None, 1)
+            if code == "+":
+                return results
+            elif code == "-":
+                raise CommunicationError(f"Arduino error: {param}")
+            elif code == ">":
+                results.append(param)
+            elif code == "#":
+                pass  # Ignore comment lines
+            else:
+                raise CommunicationError(
+                    f"Arduino returned unrecognised response line: {line}",
+                )
+
+    def _update_digital_pin(self, identifier: int) -> None:
+        assert identifier < FIRST_ANALOGUE_PIN
+        pin = self._digital_pins[identifier]
+        char: str
+        if pin.mode == GPIOPinMode.DIGITAL_INPUT:
+            char = "Z"
+        elif pin.mode == GPIOPinMode.DIGITAL_INPUT_PULLUP:
+            char = "P"
+        elif pin.mode == GPIOPinMode.DIGITAL_OUTPUT:
+            if pin.state:
+                char = "H"
+            else:
+                char = "L"
+        else:
+            assert False, "unreachable"
```

Comment:
This will be compiled out in optimised bytecode.

Reply:
For the record, this was broken out into #344.